File: golang/internal/presentation/asset_controller.go
package presentation

import (
	"errors"

	"folio/codinginterview/internal/application/usecase/asset"
)

type StockDto struct {
	Symbol           string
	EvaluationAmount string
}

type GetAssetRequest struct {
	UserId string
}

type GetAssetResponse struct {
	CashAmount string
	Stocks     []StockDto
}

type AssetController struct {
	getAssetUsecase *asset.GetAssetUsecase
}

func NewAssetController(getAssetUsecase *asset.GetAssetUsecase) *AssetController {
	return &AssetController{getAssetUsecase: getAssetUsecase}
}

func (c *AssetController) GetAsset(req GetAssetRequest) (GetAssetResponse, error) {
	uid, err := parseUserId(req.UserId)
	if err != nil {
		return GetAssetResponse{}, err
	}

	out, err := c.getAssetUsecase.Run(asset.GetAssetUsecaseInput{UserId: uid})
	if err != nil {
		if errors.Is(err, asset.ErrUserNotFound) {
			return GetAssetResponse{}, newBadRequest("user not found")
		}
		return GetAssetResponse{}, err
	}

	stocks := make([]StockDto, 0, len(out.Stocks))
	for _, s := range out.Stocks {
		stocks = append(stocks, StockDto{
			Symbol:           string(s.Symbol),
			EvaluationAmount: s.EvaluationAmount.String(),
		})
	}
	return GetAssetResponse{
		CashAmount: out.CashAmount.String(),
		Stocks:     stocks,
	}, nil
}

File: scala/src/main/scala/folio/codinginterview/presentation/AssetController.scala
package folio.codinginterview.presentation

import folio.codinginterview.application.usecase.asset.GetAssetUsecase
import folio.codinginterview.application.usecase.asset.GetAssetUsecaseException
import folio.codinginterview.application.usecase.asset.GetAssetUsecaseInput
import folio.codinginterview.presentation.PresentationException.BadRequestException
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

object AssetController {
  final case class StockDto(symbol: String, evaluationAmount: String)
  final case class GetAssetRequest(userId: String)
  final case class GetAssetResponse(cashAmount: String, stocks: Seq[StockDto])
}

final class AssetController(
    getAssetUsecase: GetAssetUsecase
)(using ec: ExecutionContext)
    extends PresentationPreparation {
  import AssetController.*

  def getAsset(req: GetAssetRequest): Future[GetAssetResponse] =
    for {
      uid <- parseUserId(req.userId)
      out <- getAssetUsecase.run(GetAssetUsecaseInput(uid)).recoverWith { case GetAssetUsecaseException.UserNotFound =>
        Future.failed(BadRequestException("user not found"))
      }
    } yield GetAssetResponse(
      cashAmount = out.cashAmount.toString,
      stocks = out.stocks.map(e => StockDto(e.symbol.toString, e.evaluationAmount.toString))
    )
}
